hypothesis-python/src/hypothesis/internal/constants_ast.py

```diff
@@ -0,0 +1,129 @@
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Copyright the Hypothesis Authors.
+# Individual contributors are listed in AUTHORS.rst and the git log.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+import ast
+import inspect
+import math
+import sys
+from ast import AST, Constant, NodeVisitor, UnaryOp, USub
+from functools import lru_cache
+from types import ModuleType
+from typing import TYPE_CHECKING, Optional, Union
+
+from hypothesis.internal.escalation import is_hypothesis_file
+from hypothesis.internal.scrutineer import ModuleLocation
+
+if TYPE_CHECKING:
+    from typing import TypeAlias
+
+ConstantT: "TypeAlias" = Union[int, float, bool, bytes, str]
+
+
+class ConstantVisitor(NodeVisitor):
+    def __init__(self):
+        super().__init__()
+        self.constants: set[ConstantT] = set()
+
+    def _add_constant(self, constant):
+        self.constants |= self._unfold_constant(constant)
+
+    # the code `a = -1` is actually a combination of a USub unary op, and the
+    # constant 1.
+    def visit_UnaryOp(self, node):
+        assert isinstance(node, UnaryOp)
+        if isinstance(node.op, USub) and isinstance(node.operand, Constant):
+            self._add_constant(-1 * node.operand.value)
+            # don't recurse on this node, or else we would add the positive variant
+            return
+
+        self.generic_visit(node)
+
+    def visit_JoinedStr(self, node):
+        # dont recurse on JoinedStr, i.e. f strings. Constants that appear *only*
+        # in f strings are unlikely to be helpful.
+        return
+
+    @classmethod
+    def _unfold_constant(cls, value: object) -> set[ConstantT]:
+        if isinstance(value, str) and (
+            len(value) > 20 or value.isspace() or value == ""
+        ):
+            # discard long strings, which are likely to be docstrings.
+            # TODO we should always ignore strings directly after a FunctionDef
+            # node, regardless of length
+            return set()
```

Comment:
Hmm, we could check whether the `ast.Expr` contains only a constant string, and skip those?  That's docstrings in general - module, function, and attribute; fiddly to express but seems doable.

Reply:
Beautiful! And pretty easy to express.